src/nodes/NodeConnectionManager.ts

```diff
@@ -0,0 +1,706 @@
+   *
+   * @param relayNodeId node ID of the relay node (i.e. the seed node)
+   * @param sourceNodeId node ID of the current node (i.e. the sender)
+   * @param targetNodeId node ID of the target node to hole punch
+   * @param egressAddress stringified address of `egressHost:egressPort`
+   * @param signature signature to verify source node is sender (signature based
+   * on egressAddress as message)
+   */
+  @ready(new nodesErrors.ErrorNodeConnectionManagerNotRunning())
+  public async sendHolePunchMessage(
+    relayNodeId: NodeId,
+    sourceNodeId: NodeId,
+    targetNodeId: NodeId,
+    egressAddress: string,
+    signature: Buffer,
+  ): Promise<void> {
+    const relayMsg = new nodesPB.Relay();
+    relayMsg.setSrcId(nodesUtils.encodeNodeId(sourceNodeId));
+    relayMsg.setTargetId(nodesUtils.encodeNodeId(targetNodeId));
+    relayMsg.setEgressAddress(egressAddress);
+    relayMsg.setSignature(signature.toString());
+    await this.withConnF(relayNodeId, async (connection) => {
+      const client = connection.getClient();
+      await client.nodesHolePunchMessageSend(relayMsg);
+    });
+  }
+
+  /**
+   * Forwards a received hole punch message on to the target.
+   * If not known, the node ID -> address mapping is attempted to be discovered
+   * through Kademlia (note, however, this is currently only called by a 'broker'
+   * node).
+   * @param message the original relay message (assumed to be created in
+   * nodeConnection.start())
+   */
+  @ready(new nodesErrors.ErrorNodeConnectionManagerNotRunning())
+  public async relayHolePunchMessage(message: nodesPB.Relay): Promise<void> {
+    await this.sendHolePunchMessage(
+      validationUtils.parseNodeId(message.getTargetId()),
+      validationUtils.parseNodeId(message.getSrcId()),
+      validationUtils.parseNodeId(message.getTargetId()),
+      message.getEgressAddress(),
+      Buffer.from(message.getSignature()),
+    );
+  }
+
+  /**
+   * Returns an array of the seed nodes.
+   */
+  @ready(new nodesErrors.ErrorNodeConnectionManagerNotRunning())
+  public getSeedNodes(): NodeId[] {
+    const seedNodes: NodeId[] = [];
+    for (const seedNodeId in this.seedNodes) {
+      seedNodes.push(IdInternal.fromString<NodeId>(seedNodeId));
+    }
+    return seedNodes;
+  }
+
+  /**
+   * This creates an interface for the NodeNodeConnection to interact with the NodeConnectionManager.
+   * @param nodeId The nodeId the NodeConnection is connected to.
+   */
+  protected getInterface(nodeId: NodeId): NodeConnectionManagerInterface {
+    return {
+      withConnF: async (...args) => {
+        return this.withConnF(...args);
+      },
+      sendHolePunchMessage: async (...args) => {
+        return this.sendHolePunchMessage(...args);
+      },
+      getSeedNodes: () => {
+        return this.getSeedNodes();
+      },
+      killSelf: async () => {
+        await this.destroyConnection(nodeId);
+      },
+    };
+  }
+}
```

Comment:
Don't think this is necessary. It should suffice to just pass `this` in, and rely on TS type checking to prevent using the wrong functions.

Only issue is the `killSelf` thing. I'm pretty sure this can be done in a better way like how network does it with `ConnectionForward` and `ConnectionReverse`.

Reply:
The interface was to limit what the `NodeConnection` could access. I was thinking about the rule of least privilege here.

Reply:
The interface is fine, but I'm talking about creating a synthetic proxy object for it. No need when we can rely on TS type checking.

Reply:
True, but we would have to replace `killSelf` with just `destroyConnection`

Reply:
> Everyonce and a while I see these weird logging messages:
> 
> ```
>     Failed to connect to 127.0.0.1:37445/?nodeId=vi3et1hrpv2m2lrplcm7cu913kr45v51cak54vm68anlbvuf83ra0 through proxy 127.0.0.1:38231 with status 526
> ```
> 
> These are coming from upstream `grpc-js` library, and that's because `node_modules/@grpc/grpc-js/src/logging.ts` uses the its own logger, and is affected by the logging verbosity.
> 
> It's a bit annoying when log messages are meant to be going to our logger system.
> 
> However I found out that it is possible to set that logger: https://github.com/grpc/grpc-node/blob/a0f298c514c6ac56e33c149f8477d35deecbb4d2/packages/grpc-js/src/logging.ts#L59-L61
> 
> This would be done like this:
> 
> ```ts
> import Logger from '@matrixai/logger';
> import * as grpc from '@grpc/grpc-js';
> 
> grpc.setLogger(new Logger('grpc') as unknown as Partial<Console>);
> ```
> 
> Some issues with this:
> 
> 1. This is a module-level property, which means it is set once globally for all grpc imports. Where would we do this? It would have to be done everywhere GRPC is imported. If we do it once at the entry-point of our program, like `index.ts`, but this wouldn't affect anything that doesn't end up using the `index.ts`, and we would to affect our tests as well.
> 2. The logger object it expects needs to match this API:
>    ```ts
>    const DEFAULT_LOGGER: Partial<Console> = {
>      error: (message?: any, ...optionalParams: any[]) => {
>        console.error('E ' + message, ...optionalParams);
>      },
>      info: (message?: any, ...optionalParams: any[]) => {
>        console.error('I ' + message, ...optionalParams);
>      },
>      debug: (message?: any, ...optionalParams: any[]) => {
>        console.error('D ' + message, ...optionalParams);
>      },
>    }
>    ```
> 3. While our `Logger` does support `error`, `info` and `debug`, it doesn't have all the other optional parameters. To do so we would have to adapt it. By creating a proxy object before calling `logger`.
>    ```ts
>    grpc.setLogger(
>      {
>        error: (...data: Array<any>) => logger.error(data.map(d => d.toString()).join(' ')),
>        info: (...data: Array<any>) => logger.info(data.map(d => d.toString()).join(' ')),
>        debug: (...data: Array<any>) => logger.debug(data.map(d => d.toString()).join(' ')),
>      }
>    )
>    ```
> 4. We would also need to convert our logging verbosity to `grpc.setLogVerbosity(grpc.logVerbosity.INFO)`... etc.

This is now used in `src/bin/polykey-agent.ts` and `src/bin/CommandPolykey.ts`. These are 2 places where they are global entrypoints to the PK program, and the GRPC upstream logger is now set appropriately. However this won't affect testing. So when we are testing things that may involve GRPC. we should manually do this:

```ts
import grpcSetLogger from '@/grpc/utils/setLogger';

describe('...', () => {
  const logger = new Logger('NodeGraph Test', LogLevel.ERROR, [
    new StreamHandler(),
  ]);
  grpcSetLogger(logger.getChild('grpc'));
});
```

@joshuakarp @tegefaulkes @emmacasolin - this applies to any test using node connection, grpc client, grpc server... etc.

Reply:
If we remove the killSelf ( now called `destroyCallback`) then we can remove this function can cast the interface directly. the `destroyCallback` will need to be created separately.

Reply:
I've removed the `getInterface` method. the `NodeConnectionManagerInterface` type was moved to `NodeConnection.ts`. `destroyCallback` has been seperated out of the interface and passed separately to `NodeConnection`.